src/pip/_internal/resolution/resolvelib/found_candidates.py

```diff
@@ -0,0 +1,134 @@
+    from typing import Callable, Iterator, Optional, Set
+
+    from pip._vendor.packaging.version import _BaseVersion
+
+    from .base import Candidate
+
+
+class _InstalledFirstCandidatesIterator(collections_abc.Iterator):
+    """Iterator for ``FoundCandidates``.
+
+    This iterator is used when the resolver prefers to keep the version of an
+    already-installed package. The already-installed candidate is always
+    returned first. Candidates from index are accessed only when the resolver
+    wants them, and the already-installed version is excluded from them.
+    """
+    def __init__(
+        self,
+        get_others,  # type: Callable[[], Iterator[Candidate]]
+        installed,  # type: Optional[Candidate]
+    ):
+        self._installed = installed
+        self._get_others = get_others
+        self._others = None  # type: Optional[Iterator[Candidate]]
+        self._returned = set()  # type: Set[_BaseVersion]
+
+    def __next__(self):
+        # type: () -> Candidate
+        if self._installed and self._installed.version not in self._returned:
+            self._returned.add(self._installed.version)
+            return self._installed
+        if self._others is None:
+            self._others = self._get_others()
+        cand = next(self._others)
+        while cand.version in self._returned:
+            cand = next(self._others)
+        self._returned.add(cand.version)
+        return cand
+
+    next = __next__  # XXX: Python 2.
+
+
+class _InstalledReplacesCandidatesIterator(collections_abc.Iterator):
+    """Iterator for ``FoundCandidates``.
+
+    This iterator is used when the resolver prefers to upgrade an
+    already-installed package. Candidates from index are returned in their
+    normal ordering, except replaced when the version is already installed.
+    """
+    def __init__(
+        self,
+        get_others,  # type: Callable[[], Iterator[Candidate]]
+        installed,  # type: Optional[Candidate]
+    ):
+        self._installed = installed
+        self._get_others = get_others
+        self._others = None  # type: Optional[Iterator[Candidate]]
+        self._returned = set()  # type: Set[_BaseVersion]
+
+    def __next__(self):
+        # type: () -> Candidate
+        if self._others is None:
+            self._others = self._get_others()
+        try:
+            cand = next(self._others)
+            while cand.version in self._returned:
+                cand = next(self._others)
+            if self._installed and cand.version == self._installed.version:
+                cand = self._installed
+        except StopIteration:
+            # Return the already-installed candidate as the last item if its
+            # version does not exist on the index.
+            if not self._installed:
+                raise
+            if self._installed.version in self._returned:
+                raise
+            cand = self._installed
+        self._returned.add(cand.version)
+        return cand
```

Comment:
Doesn't this handle development versions differently compared to status quo (AKA old resolver)? I guess we implemented the same thing in current master, so.. :shrug:.

Reply:
Yup -- there's a change in behavior here.
```
❯ pip install -U pip        
Requirement already up-to-date: pip in ./lol/lib/python3.8/site-packages (20.3.dev0)
❯ pip install -U pip --use-feature=2020-resolver
Requirement already satisfied: pip in ./lol/lib/python3.8/site-packages (20.3.dev0)

❯ g co-pr upstream 8932
❯ pip install ~/Projects/pip

❯ pip install -U pip
Requirement already up-to-date: pip in ./lol/lib/python3.8/site-packages (20.3.dev0)
❯ pip install -U pip --use-feature=2020-resolver
Requirement already satisfied: pip in ./lol/lib/python3.8/site-packages (20.3.dev0)
Collecting pip
  Using cached pip-20.2.3-py2.py3-none-any.whl (1.5 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 20.3.dev0
    Uninstalling pip-20.3.dev0:
      Successfully uninstalled pip-20.3.dev0
Successfully installed pip-20.2.3
```